docs/architecture-decision-records/006-stage-awareness.md

```diff
@@ -0,0 +1,237 @@
+  - Stage dependent values will be placed in the [CDK context][cdk-context], which is good
+  - This solution would likely require our build artifacts follow the [Cloud Assembly spec][cloud-assembly]. This is a departure from the status quo and a big, possibly breaking, change.
+- :white_check_mark: How short the feedback loop is
+  - We'd be using `cdk` exactly how AWS suggest, using all their tooling which is quite mature now with good DX
+- :x: Ease of maintenance
+  - This solution would require significant changes to Riff-Raff
+  - The `cdk` toolkit is only available as a CLI tool, and there are no known plans for a Java SDK to be supported
+
+### 4. Produce a CloudFormation template for each stage
+In this model, we step away from the "best practice" of using the exact same artifact between stages.
+However, the artifact will be produced by the same CDK pipeline, and thus we retain the deterministic characteristic.
+
+For a stack to gain stage-awareness, one would simply instantiate the same class:
+
+```ts
+class MyStack extends GuStack { } // `GuStack` is provided by GuCDK
+
+new MyStack(scope, "MyStack", {
+ stack: "the-stack",
+ stage: "PROD",
+ minCapacity: 3,
+ domainName: "example.com"
+});
+
+new MyStack(scope, "MyStack", {
+ stack: "the-stack",
+ stage: "CODE",
+ minCapacity: 1,
+ domainName: "code.example.com"
+});
+```
+#### Score
+- :white_check_mark: Adherence to the best practice of producing a single, deterministic, artifact
+  - Whilst not a _single_ artifact, it is deterministic
+  - Unit or snapshot tests can be used to ensure the resulting template across stages only differ in property values and not the overall infrastructure
+- :x: (but possible :white_check_mark:) Adherence to the best practice of continuous delivery of infrastructure
+  - Riff-Raff does not currently support the use of a different file per stage
+- :white_check_mark: Support stacks that do not use `CODE` or `PROD`
+- :white_check_mark: The simplicity of the API presented to users
+  - The user defines the shape of input to `MyStack` and can use _any_ type, including strong types such as [`InstanceType`][cdk-instance-type]
+- :white_check_mark: How short the feedback loop is
+  - Standard `cdk` workflows can be employed
+- :white_check_mark: Ease of maintenance
+  - Inline with [AWS CDK recommendations][cdk-recommendations]
+  - A small change to Riff-Raff would be needed. However, this it should be possible to do this in a non-breaking way.
+
+## Decision
+
+Produce a CloudFormation template for each stage.
+
+## Consequences
+
+Riff-Raff would require some changes to enable the use of a different CloudFormation template file depending on the stage being deployed.
+
+The `riff-raff.yaml` would likely look like this:
+
+```yaml
+stacks:
+  - playground
+regions:
+  - eu-west-1
+deployments:
+  cloudformation:
+    type: cloud-formation
+    app: cdk-playground
+    parameters:
+      templateStagePath: # <- NEW!
```

Comment:
Might be worth citing the existing stage-specific parameter that already exists here. I.e. this kind of approach is not entirely new.